### PR TITLE
additional org apis

### PR DESCRIPTION
**Pull request by fillup**

### CODE REVIEW

File: application/actions/organization_test.go [hunk 1]
```diff
@@ -4,11 +4,13 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
-	"io/ioutil"
```
Comment (Schparky) [inline application/actions/organization_test.go]: I like to keep the internal packages grouped at the top.

File: application/actions/organization_test.go [hunk 2]
```diff
@@ -229,3 +231,156 @@ func (as *ActionSuite) Test_OrganizationCreateRemoveUpdate() {
...
+func (as *ActionSuite) Test_OrganizationViewAndList() {
+	t := as.T()
+
+	// create organizations
+	org1 := models.Organization{
+		Name:                "Org1",
+		Url:                 nulls.String{},
+		AuthType:            models.AuthTypeSaml,
+		AuthConfig:          "{}",
+		UUID:                domain.GetUUID(),
```
Comment (Schparky) [inline application/actions/organization_test.go]: `MustCreate` assigns a new UUID if a `uuid` field exists
```suggestion
```

File: application/actions/organization_test.go [hunk 3]
```diff
@@ -229,3 +231,156 @@ func (as *ActionSuite) Test_OrganizationCreateRemoveUpdate() {
...
+	}
+	test.MustCreate(as.DB, &org1)
+	// create organizations
+	org2 := models.Organization{
+		Name:                "Org2",
+		Url:                 nulls.String{},
+		AuthType:            models.AuthTypeSaml,
+		AuthConfig:          "{}",
+		UUID:                domain.GetUUID(),
```
Comment (Schparky) [inline application/actions/organization_test.go]:
```suggestion
```

File: application/actions/organization_test.go [hunk 4]
```diff
@@ -229,3 +231,156 @@ func (as *ActionSuite) Test_OrganizationCreateRemoveUpdate() {
...
+		resp := httptest.NewRecorder()
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", tc.Token))
+		req.Header.Set("content-type", "application/json")
+
+		as.App.ServeHTTP(resp, req)
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			t.Error(err)
+		}
```
Comment (Schparky) [inline application/actions/organization_test.go]: This test is a little different, but just wanted to highlight `actions.testGqlQuery` that does all of this and more.

File: application/go.mod [hunk 5]
```diff
@@ -1,6 +1,6 @@
 module github.com/silinternational/wecarry-api
 
-go 1.12
+go 1.13
```
Comment (Schparky) [inline application/go.mod]: Should `Dockerfile` and `Dockerfile-dev` be updated as well?

File: application/actions/organization_test.go (hunk 1 above)
Comment (fillup) [reply application/actions/organization_test.go]: that must have been done automatically by GoLand

File: application/actions/organization_test.go (hunk 4 above)
Comment (fillup) [reply application/actions/organization_test.go]: I saw that but didn't look into exactly how it worked and I just used the same process/convention I had for a previous test. should I refactor these to use the shared one?

File: application/go.mod (hunk 5 above)
Comment (fillup) [reply application/go.mod]: good point, I guess so

File: application/actions/organization_test.go (hunk 4 above)
Comment (Schparky) [reply application/actions/organization_test.go]: It's up to you. The other method parses the received data into a struct so you can compare field values.

Comment (Schparky) [reply application/actions/organization_test.go]: Yours is better set up for multiple tests/scenarios, though.

File: application/models/user.go [hunk 6]
```diff
@@ -243,7 +243,34 @@ func (u *User) CanCreateOrganization() bool {
...
+// CanViewOrganization returns true if the given user is allowed to view the specified organization
+func (u *User) CanViewOrganization(orgId int) bool {
+	// if user is a system admin, allow
+	if u.AdminRole == UserAdminRoleSuperAdmin || u.AdminRole == UserAdminRoleSalesAdmin {
+		return true
+	}
+
+	// make sure we're checking current user orgs
+	err := DB.Load(u, "UserOrganizations")
+	if err != nil {
```
Comment (Baggerone) [inline application/models/user.go]: You could combine these two lines.

File: application/gqlgen/organization.go [hunk 7]
```diff
@@ -13,6 +15,43 @@ func (r *Resolver) Organization() OrganizationResolver {
...
+func (r *queryResolver) Organizations(ctx context.Context) ([]models.Organization, error) {
+	cUser := models.GetCurrentUserFromGqlContext(ctx)
+	extras := map[string]interface{}{
+		"user": cUser.UUID,
+	}
+
+	// get list of orgs that cUser is allowed to see
+	orgs := models.Organizations{}
+	err := orgs.AllWhereUserIsOrgAdmin(cUser)
+	if err != nil {
```
Comment (Baggerone) [inline application/gqlgen/organization.go]: You could combine these two lines.

File: application/models/user.go (hunk 6 above)
Comment (fillup) [reply application/models/user.go]: sure, will fix

File: application/gqlgen/organization.go (hunk 7 above)
Comment (fillup) [reply application/gqlgen/organization.go]: sure

File: application/gqlgen/organization.go [hunk 8]
```diff
@@ -13,6 +15,43 @@ func (r *Resolver) Organization() OrganizationResolver {
...
+func (r *queryResolver) Organization(ctx context.Context, id *string) (*models.Organization, error) {
+	cUser := models.GetCurrentUserFromGqlContext(ctx)
+	extras := map[string]interface{}{
+		"user":    cUser.UUID,
+		"orgUUID": *id,
+	}
+
+	org := &models.Organization{}
+	err := org.FindByUUID(*id)
+	if domain.IsOtherThanNoRows(err) {
```
[the remark this answers is not in the document]
Comment (fillup) [reply application/gqlgen/organization.go]: ok

File: application/actions/organization_test.go [hunk 9]
```diff
@@ -229,3 +231,156 @@ func (as *ActionSuite) Test_OrganizationCreateRemoveUpdate() {
...
+
+		for _, sub := range tc.ExpectSubStrings {
+			if !strings.Contains(string(body), sub) {
+				t.Errorf("ExpectSubString \"%s\" not found in response: %s. test case: %s", sub, string(body), tc.Name)
+			}
+		}
+
+		for _, sub := range tc.DontExpectSubStrings {
+			if strings.Contains(string(body), sub) {
+				t.Errorf("DontExpectSubString \"%s\" found in response and should not have been: %s. test case: %s", sub, string(body), tc.Name)
```
[the remark this answers is not in the document]
Comment (fillup) [reply application/actions/organization_test.go]: I rewrote all that, so it's using as.Equal now

File: application/actions/organization_test.go [hunk 10]
```diff
@@ -4,11 +4,13 @@ import (
...
 	"fmt"
-	"io/ioutil"
-	"strings"
-
 	"github.com/gobuffalo/httptest"
+	"github.com/gobuffalo/nulls"
+	"github.com/silinternational/wecarry-api/domain"
+	"github.com/silinternational/wecarry-api/internal/test"
 	"github.com/silinternational/wecarry-api/models"
+	"io/ioutil"
```
[the remark this answers is not in the document]
Comment (fillup) [reply application/actions/organization_test.go]: ok I fixed my settings, I'd trashed the .idea folder when trying to fix that other issue a few days ago so I lost some project prefs like `go fmt` too, but that's back in place

**Pull request merged**